autoPyTorch/pipeline/components/training/trainer/base_trainer.py

```diff
@@ -0,0 +1,351 @@
+                    epoch * len(train_loader) + step,
+                )
+
+        return loss_sum / N, self.compute_metrics(outputs_data, targets_data)
+
+    def evaluate(self, test_loader: torch.utils.data.DataLoader, epoch: int
+                 ) -> Tuple[float, Dict[str, float]]:
+        '''
+            Evaluates the model in both metrics and criterion
+
+        Args:
+            test_loader (torch.utils.data.DataLoader): generator of features/label
+            epoch (int): the current epoch for tracking purposes
+
+        Returns:
+            float: test loss
+            Dict[str, float]: scores for each desired metric
+        '''
+        self.model.eval()
+
+        loss_sum = 0.0
+        N = 0
+        outputs_data = list()
+        targets_data = list()
+
+        with torch.no_grad():
+            for step, (data, targets) in enumerate(test_loader):
+
+                batch_size = data.shape[0]
+                data = data.float().to(self.device)
+                targets = targets.long().to(self.device)
+                outputs = self.model(data)
+                loss = self.criterion(outputs, targets)
+                loss_sum += loss.item() * batch_size
+                N += batch_size
+
+                outputs_data.append(outputs.detach())
+                targets_data.append(targets.detach())
+
+                if self.writer:
+                    self.writer.add_scalar(
+                        'Val/loss',
+                        loss.item(),
+                        epoch * len(test_loader) + step,
+                    )
+
+        self.model.train()
+        return loss_sum / N, self.compute_metrics(outputs_data, targets_data)
+
+    def compute_metrics(self, outputs_data: np.ndarray, targets_data: np.ndarray
+                        ) -> Dict[str, float]:
+        # TODO: change once Ravin Provides the PR
+        outputs_data = torch.cat(outputs_data, dim=0)
+        targets_data = torch.cat(targets_data, dim=0)
+        return {
+            metric.get_properties()['name']: metric(
+                outputs_data, targets_data) for metric in self.metrics
+        }
+
+    def data_preparation(self, X: np.ndarray, y: np.ndarray,
+                         ) -> Tuple[np.ndarray, Dict[str, np.ndarray]]:
+        """
+        Depending on the trainer choice, data fed to the network might be pre-processed
+        on a different way. That is, in standard training we provide the data to the
+        network as we receive it to the loader. Some regularization techniques, like mixup
+        alter the data.
+
+        Args:
+            X (np.ndarray): The batch training features
+            y (np.ndarray): The batch training labels
+
+        Returns:
+            np.ndarray: that processes data
+            Dict[str, np.ndarray]: arguments to the criterion function
+        """
+        raise NotImplementedError()
+
+    def criterion_preparation(self, y_a: np.ndarray, y_b: np.ndarray = None, lam: float = 1.0
```

Comment:
I assume y_a and y_b are used for mixup. Maybe it would be cleaner to just have y here and in the StandardTrainer and have the MixupTrainer override this method

Reply:
This is the one comment I cannot address. Mypy does not allow this redefinition as we need to comply with the parent-child implementation. I used to have it as you proposed (because autopytorch originally had it like this) But with mypy, we need to comply with the parent class interface

Reply:
Okay, then let's just go with this. This allows us to introduce weights via y_b which is hopefully enough for most cases